client/comms/wsconn.go

```diff
@@ -0,0 +1,349 @@
+		return err
+	}
+
+	conn.setConnected(true)
+	conn.ws.SetPingHandler(func(string) error {
+		now := time.Now()
+		err := conn.ws.SetReadDeadline(now.Add(conn.cfg.PingWait))
+		if err != nil {
+			log.Errorf("unable to set read deadline: %v", err)
+			return err
+		}
+
+		// Respond with a pong.
+		return conn.ws.WriteControl(websocket.PongMessage, []byte{},
+			now.Add(conn.cfg.WriteWait))
+	})
+
+	return nil
+}
+
+// keepAlive maintains an active websocket connection by reconnecting when
+// the established connection is broken. This should be run as a goroutine.
+func (conn *WsConn) keepAlive(ctx context.Context) {
+	ticker := time.NewTicker(time.Second)
+
+	for {
+		select {
+		case <-ticker.C:
+			if !conn.isConnected() {
+				err := conn.connect()
+				if err != nil {
+					log.Errorf("unable to establish a connection: %v", err)
+				}
+
+				// Run the reconnect sync after reestablishing a connection.
+				if conn.cfg.ReconnectSync != nil {
+					conn.cfg.ReconnectSync()
+				}
+			}
+
+		case <-ctx.Done():
+			now := time.Now()
+			closeMsg := websocket.FormatCloseMessage(websocket.CloseMessage, "")
+			conn.writeMtx.Lock()
+			err := conn.ws.WriteControl(websocket.CloseMessage, closeMsg,
+				now.Add(conn.cfg.WriteWait))
+			conn.writeMtx.Unlock()
+			if err != nil {
+				log.Errorf("unable to close websocket connection: %v", err)
+			}
+			ticker.Stop()
+			conn.wg.Done()
+			return
+		}
+	}
+}
+
+// read fetches and parses incoming messages for processing. This should be
+// run as a goroutine.
+func (conn *WsConn) read() {
+	for {
+		// Only allow reads if there is an established connection.
+		if !conn.isConnected() {
+			time.Sleep(time.Millisecond * 500)
+			continue
+		}
+
+		msg := new(msgjson.Message)
+		err := conn.ws.ReadJSON(msg)
+		if err != nil {
+			if opErr, ok := err.(*net.OpError); ok {
+				if opErr.Op == "read" {
+					conn.setConnected(false)
+					continue
+				}
+			}
+
+			if websocket.IsCloseError(err, websocket.CloseAbnormalClosure,
+				websocket.CloseGoingAway, websocket.CloseNormalClosure) {
+				return
+			}
+
+			if nErr, ok := err.(net.Error); ok {
+				if nErr.Timeout() {
+					conn.setConnected(false)
+					continue
+				}
+			}
+
+			if _, ok := err.(*json.UnmarshalTypeError); ok {
+				log.Error(err)
+				continue
+			}
+
+			log.Errorf("unable to read message: %v", err)
+			conn.cfg.Cancel()
+			return
+		}
+
+		conn.readCh <- msg
+	}
+}
+
+// send pushes outgoing messages over the websocket connection.
+// This should be run as a goroutine.
+func (conn *WsConn) send(ctx context.Context) {
+	for {
+		select {
+		case msg := <-conn.sendCh:
+			// Log the route if the message being sent is a request.
+			if msg.Type == msgjson.Request {
+				conn.logRequest(msg.ID, msg)
+			}
+
+			conn.writeMtx.Lock()
+			_ = conn.ws.SetWriteDeadline(time.Now().Add(conn.cfg.WriteWait))
+			err := conn.ws.WriteJSON(msg)
+			conn.writeMtx.Unlock()
+			if err != nil {
+				log.Errorf("unable to write message: %v", err)
```

Comment:
When/why is it ok to continue?  Please make a comment in the code addressing the question if this is intended.

Reply:
It's ok to continue for json parsing errors because the underlying connection is still valid and can push valid messages sent. It's ok to continue on read and timeout errors because there is a `keepAlive` goroutine that will reestablish a broken connection. It's also ok to continue on write errors as well due to the keep alive process. The only cases I can see where it's not ok to continue is if a normal or abnormal close message is received.

Reply:
Yeah, that sounds good.  I'm just asking for the reasoning to be put into the comments. Documenting the intent is critical to understand when something is working properly (someone will come along in some years and think they know better, change it, break it).

Reply:
My bad, sorry about that.